src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubLeaseFactory.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.ServiceBus
+{
+    // EventProcessorHost requires a container name that the multiple host instances use to coordinate with each other and create leases. 
+    // An event hub is described by an (namespace name, event hub path name). The mapping from the event hub to the container name
+    // must be a one-to-one injective function such that:
+    // 1. it's deterministic: when we scale out ot N separate instances, they must all compute the same container name for the same event hub. 
+    // 2. it's collision free: different event hubs shouldn't share the same container. So we can't just use a hash. 
+    //  Note that the event hub name is much more expressive than the storage container name, so we can't do a stateless collision-free transform. 
+    // Hence we use a lookaside storage to save the transform. 
+    internal class EventHubLeaseFactory
+    {
+        // Name of Azure blob that maps "eventHub" --> container name. 
+        private const string InternalContainerName = "azure-webjobs-hosts";
+        private const string InternalBlobPath = "eventhub/leasenames.json";
+
+        // Concurrent safe mutation of a blob. 
+        // Apply mutator to update the blob (as a deserialized JSON contents). 
+        // If blob doesn't exist yet, then this allocates a new TBlob() object and calls the mutator on that. 
+        // Mutator also extracts a value and returns it. 
+        private static TResult MutateBlob<TBlob, TResult>(CloudBlockBlob blob, Func<TBlob, TResult> mutator) where TBlob : new()
```

Comment:
Mutating a single blob seems a lot more complicated and error prone than simply doing:

- GET azure-webjobs-hosts/eventhub/lease-containers/<namespace>-<hubname>
- If blob exists, read it to get the container name, else PUT a new blob to that path with the container name

That's basically the same logic you had before for table storage, except instead of table queries you're doing direct blob references.